internal/dataplane/kongstate/util.go

```diff
@@ -16,13 +16,45 @@ import (
 	configurationv1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1"
 )
 
-func getKongIngressForService(s store.Storer, service corev1.Service) (
-	*configurationv1.KongIngress, error) {
-	confName := annotations.ExtractConfigurationName(service.Annotations)
-	if confName == "" {
-		return nil, nil
+func getKongIngressForServices(
+	s store.Storer,
+	services map[string]*corev1.Service,
+) (*configurationv1.KongIngress, error) {
+	// loop through each service and retrieve the attached KongIngress resources.
+	// there can only be one KongIngress for a group of services: either one of
+	// them is configured with a KongIngress and this configures the Kong Service
+	// or Upstream OR all of them can be configured but they must be configured
+	// with the same KongIngress.
+	var kongIngress *configurationv1.KongIngress
+	for _, svc := range services {
+		// check if the service is even configured with a KongIngress
+		confName := annotations.ExtractConfigurationName(svc.Annotations)
+		if confName == "" {
+			continue // some other service in the group may yet have a KongIngress attachment
+		}
+
+		// retrieve the attached KongIngress for the service
+		newKongIngress, err := s.GetKongIngress(svc.Namespace, confName)
+		if err != nil {
+			return nil, err
+		}
+
+		// if we've already seen a KongIngress for a previous Kubernetes service
+		// this is only allowed for a group of services if the KongIngress is the
+		// same for each Service.
```

Comment:
Split between annotation equality and KongIngress equality in kongstate and parser/ingressrules seems a bit odd. Do we need to do both? KongIngress is added via the `override` annotation, so do we not get KongIngress consistency by virtue of checking override annotation consistency?

Reply:
You're right that we most likely can't hit it. The main purpose of this function was to make it possible to set a `KongIngress` on even just 1 service in the backends to enact the overrides, and thus I felt like a safeguard might be good there for the future. In retrospect however, maybe I shouldn't try to allow this. Maybe I should simply require that all `Services` have the _exact same_ overriding `KongIngress` instead, wdyt?

Reply:
IMO we shouldn't allow it to be set on only one: affirmative consistency is better than assumed if not explicitly contradicted. Lack of a KongIngress normally implies using the default values for anything it controls.

That poses a problem for Routes that reference Services across namespaces though, since you can't reference a KongIngress in another namespace. The same applies to (non-cluster) plugins and client certificates. Personally I'd just say we impose a restriction that says if you want to use multiple Services they must all reside in the same namespace (even if that's not the same namespace as the Route). If we do want to support multiple backends residing in different namespaces, this would be the way to approach it barring some new ClusterKongIngress.

Client certificates you're basically screwed without some check on equality of the values within the Secrets.